ocis/pkg/init/init.go

```diff
@@ -0,0 +1,306 @@
+}
+
+type AuthProviderSettings struct {
+	Oidc InsecureExtension
+}
+type AuthbearerExtension struct {
+	AuthProviders AuthProviderSettings `yaml:"auth_providers"`
+}
+
+type UserAndGroupExtension struct {
+	Drivers LdapBasedExtension
+}
+
+type ThumbnailSettings struct {
+	WebdavAllowInsecure bool `yaml:"webdav_allow_insecure"`
+	Cs3AllowInsecure    bool `yaml:"cs3_allow_insecure"`
+}
+
+type ThumbNailExtension struct {
+	Thumbnail ThumbnailSettings
+}
+
+// TODO: use the oCIS config struct instead of this custom struct
+// We can't use it right now, because it would need  "omitempty" on
+// all elements, in order to produce a slim config file with `ocis init`.
+// We can't just add these "omitempty" tags, since we want to generate
+// full example configuration files with that struct, too.
+// Proposed solution to  get rid of this temporary solution:
+// - use the oCIS config struct
+// - set the needed values like below
+// - marshal it to yaml
+// - unmarshal it into yaml.Node
+// - recurse through the nodes and delete empty / default ones
+// - marshal it to yaml
+type OcisConfig struct {
+	TokenManager      TokenManager `yaml:"token_manager"`
+	MachineAuthApiKey string       `yaml:"machine_auth_api_key"`
+	TransferSecret    string       `yaml:"transfer_secret"`
+	Graph             GraphExtension
+	Idp               LdapBasedExtension
+	Idm               IdmExtension
+	Proxy             InsecureProxyExtension
+	Frontend          FrontendExtension
+	AuthBasic         AuthbasicExtension  `yaml:"auth_basic"`
+	AuthBearer        AuthbearerExtension `yaml:"auth_bearer"`
+	User              UserAndGroupExtension
+	Group             UserAndGroupExtension
+	StorageMetadata   DataProviderInsecureSettings `yaml:"storage_metadata"`
+	StorageUsers      DataProviderInsecureSettings `yaml:"storage_users"`
+	Ocdav             InsecureExtension
+	Thumbnails        ThumbNailExtension
+}
+
+func checkConfigPath(configPath string) error {
+	targetPath := path.Join(configPath, configFilename)
+	if _, err := os.Stat(targetPath); err == nil {
+		return fmt.Errorf("config in %s already exists", targetPath)
+	}
+	return nil
+}
+
+func backupOcisConfigFile(configPath string) (string, error) {
+	sourceConfig := path.Join(configPath, configFilename)
+	targetBackupConfig := path.Join(configPath, configFilename+"."+time.Now().Format("2006-01-02-15-04-05")+".backup")
+	source, err := os.Open(sourceConfig)
+	if err != nil {
+		log.Fatalf("Could not read %s (%s)", sourceConfig, err)
+	}
+	defer source.Close()
+	target, err := os.Create(targetBackupConfig)
+	if err != nil {
+		log.Fatalf("Could not generate backup %s (%s)", targetBackupConfig, err)
+	}
+	defer target.Close()
+	_, err = io.Copy(target, source)
+	if err != nil {
+		log.Fatalf("Could not write backup %s (%s)", targetBackupConfig, err)
+	}
+	return targetBackupConfig, nil
+}
+
+func CreateConfig(insecure, forceOverwrite bool, configPath, adminPassword string) error {
```

Comment:
Nitpick: The exported functions in this file should have comments or be unexported :wink: